pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,175 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.devices.preprocess import null_postprocessing
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_unsupported_obs = frozenset(["PauliX", "PauliY", "Hadamard"])
+
+
+@transform
+def diagonalize_tape_measurements(tape, unsupported_obs=_default_unsupported_obs):
```

Comment:
Doesn't it make more sense to specify `supported_obs` instead of `unsupported_obs`? By default I would assume that `PauliZ` is the only supported obs, and the plugin developers would be free to specify any additional obs that the device supports?

Reply:
Second this idea.  I think we should opt in to natively supporting observables.  I also think default behavior should be "diagonalize everything".

Reply:
I would say that for any observable that the device claim to support, the device should know how to diagonalize it for measurement?

Reply:
The scope of the story is in regard to diagonalizing some subset of X, Y and Hadamard, so sticking with that for now. We can always extend it later if relevant.